samples/nrf9160/modem_shell/src/drivers/nrf91_non_offloaded_dev.c

```diff
@@ -0,0 +1,507 @@
+					return;
+				}
+			}
+			/* TODO: IPv6 */
+
+			ret = util_get_pdn_conn_dyn_params(0, &pdn_dyn_info);
+			if (ret) {
+				mosh_warn("Couldn't get PDN dyn info");
+			} else {
+				/* Set LTE link MTU also to non-offloading net if */
+				net_if_set_mtu(nrf91_non_offload_iface_data.iface,
+					       pdn_dyn_info.ipv4_mtu);
+
+#if defined(CONFIG_DNS_RESOLVER) && !defined(CONFIG_DNS_SERVER_IP_ADDRESSES)
+				/* Dynamic DNS servers and set them on zephyr stack */
+				struct dns_resolve_context *dnsctx;
+				struct sockaddr_in dns1 = {
+					.sin_family = AF_INET,
+					.sin_port = htons(53),
+					.sin_addr = pdn_dyn_info.dns_addr4_primary
+				};
+				struct sockaddr_in dns2 = {
+					.sin_family = AF_INET,
+					.sin_port = htons(53),
+					.sin_addr = pdn_dyn_info.dns_addr4_secondary
+				};
+				const struct sockaddr *dns_servers[] = {
+					(struct sockaddr *) &dns1,
+					(struct sockaddr *) &dns2,
+					NULL
+				};
+
+				dnsctx = dns_resolve_get_default();
+				ret = dns_resolve_reconfigure(dnsctx, NULL, dns_servers);
+				if (ret < 0) {
+					mosh_error("Could not set DNS servers");
+				}
+#endif
+			}
+
+			/* TODO: support for multiple net interfaces within non-offloading device,
+			 * I.e. each for created PDN context
+			 */
+		}
+	} else {
+		nrf91_non_offload_socket_close();
+	}
+}
+
+/******************************************************************************/
+
+static uint8_t *util_fake_dev_get_mac(struct nrf91_non_offload_dev_context *ctx)
+{
+	if (ctx->mac_addr[2] == 0x00) {
+		/* 00-00-5E-00-53-xx Documentation RFC 7042 */
+		ctx->mac_addr[0] = 0x00;
+		ctx->mac_addr[1] = 0x00;
+		ctx->mac_addr[2] = 0x5E;
+		ctx->mac_addr[3] = 0x00;
+		ctx->mac_addr[4] = 0x53;
+		ctx->mac_addr[5] = sys_rand32_get();
+	}
+
+	return ctx->mac_addr;
+}
+
+static void nrf91_non_offload_iface_init(struct net_if *iface)
+{
+	struct nrf91_non_offload_dev_context *ctx = net_if_get_device(iface)->data;
+	uint8_t *mac = util_fake_dev_get_mac(ctx);
+
+	ctx->iface = iface;
+	ctx->default_pdp_active = false;
+	ctx->mdm_skct_id = NO_MDM_SCKT;
+
+	/* The mac address is not really used but net if expects to find one. */
+	net_if_set_link_addr(iface, mac, 6, NET_LINK_ETHERNET);
+}
+
+static int nrf91_non_offload_init(const struct device *arg)
+{
+	int err = 0;
+
+	ARG_UNUSED(arg);
+
+	k_work_init_delayable(&mdm_socket_work, nrf91_non_offload_mdm_socket_worker);
+
+	err = link_shell_pdn_event_forward_cb_set(nrf91_non_offload_pdn_event_handler);
+	if (err) {
+		mosh_error("link_shell_pdn_event_forward_cb_set() failed, err %d\n", err);
+		return -1;
+	}
+	return 0;
+}
+
+/******************************************************************************/
+
+#define NRF91_MODEM_DATA_UL_BUFFER_SIZE 1500
+static char send_buffer[NRF91_MODEM_DATA_UL_BUFFER_SIZE];
+
+static int nrf91_non_offload_iface_send(const struct device *dev, struct net_pkt *pkt)
+{
+	int ret, data_len;
+
+	data_len = net_pkt_get_len(pkt);
+	ret = net_pkt_read(pkt, send_buffer, data_len);
```

Comment:
I think that `send_buffer` access should be protected with a mutex. In case there's no dedicated TX thread (CONFIG_NET_TC_TX_COUNT == 0), the transmission theoretically could take place from different threads.

Reply:
Good point. Mtx added.